# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [3]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [4]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [6]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [5]:
# Method 1. Use dtypes() method
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [6]:
# Method 2. Use info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [28]:
percentiles_gross_income = df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])
display(percentiles_gross_income)

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

In [9]:
percentile_75 = percentiles_gross_income['75%']
print(f"The 75th percentile is: {percentile_75}")

The 75th percentile is: 22.44525


**A percentile** is a statistical term used to measure the position of a value in a distribution of data.

*Interpretation using the sales example:*

The 75th percentile of gross income means that 75% of all invoices have a gross income less than or equal to this value, and the remaining 25% have a higher income. In other words, if we sort all invoices by their gross income, the value at the upper boundary of the third quartile represents the 75th percentile.

The 75th percentile is 22.44525, it means that 75% of purchases generated a profit of no more than 22.44525.


5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [11]:
invoice_counts = df['City'].value_counts()

print("Number of Invoices for Each City:")
print(invoice_counts)

Number of Invoices for Each City:
City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64


6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [18]:
# Method 1
invoice_counts = df['City'].value_counts(normalize=True)

print("Percentage of Invoices for Each City:")
print(invoice_percentages)

Percentage of Invoices for Each City:
City
Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: count, dtype: float64


In [19]:
# Method 2
invoice_percentages = (invoice_counts / invoice_counts.sum())*100

print("Percentage of Invoices for Each City:")
print(invoice_percentages)

Percentage of Invoices for Each City:
City
Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: proportion, dtype: float64


7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [23]:
avg_yangon = df[df['City'] == 'Yangon']['gross income'].mean()
avg_mandalay = df[df['City'] == 'Mandalay']['gross income'].mean()

print(f"Average gross income in Yangon: {avg_yangon}")
print(f"Average gross income in Mandalay: {avg_mandalay}")

Average gross income in Yangon: 14.874001470588235
Average gross income in Mandalay: 15.232024096385542


After analyzing the data, we see that Mandalay has a higher average gross income (15.23) compared to Yangon (14.87).

**Hypotheses:**

1. Mandalay has more premium-segment customers.
How to check: Analyze the distribution of purchases by product categories (Product line), review average transaction values, and total revenue per customer.
2. Different product categories with higher profitability are sold in Mandalay.
How to check: Compare the sales distribution by Product line in both cities and identify which categories generate the highest income.
3. Staff or marketing is more effective in Mandalay.
How to check: Compare the number of sales per employee (if data is available), review promotions, loyalty programs, and repeat purchase statistics.

_How these hypotheses help in decision-making:_

If at least one hypothesis is confirmed, we can:

- Adjust product assortment in cities with lower income by offering higher-margin products;
- Train staff based on best practices observed in Mandalay;
- Optimize marketing efforts by focusing on the most effective channels and categories identified in Mandalay.

This can help increase the overall profitability of the supermarket chain by understanding and leveraging city-specific customer behavior.

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [40]:
unique_margins = df['gross margin percentage'].nunique()
print(f"Unique values in 'gross margin percentage' column: {unique_margins}")

if unique_margins == 1:
    print("\nResult: All invoices have the same margin.")
else:
    print("\nResult: The margin differs for different invoices.")  

Unique values in 'gross margin percentage' column: 1

Result: All invoices have the same margin.


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [48]:
# Method 1
yangon_dist = df[df['City'] == 'Yangon']['Product line'].value_counts(normalize=True)
mandalay_dist = df[df['City'] == 'Mandalay']['Product line'].value_counts(normalize=True)

print("Percentage Distribution of Product line Categories - Yangon:")
print(yangon_dist)

print("\nPercentage Distribution of Product line Categories - Mandalay:")
print(mandalay_dist)

Percentage Distribution of Product line Categories - Yangon:
Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

Percentage Distribution of Product line Categories - Mandalay:
Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64


In [50]:
# Method 2
filter_df = df.query('City in ["Yangon", "Mandalay"]')
product_line_count = filter_df.groupby('City')['Product line'].value_counts(normalize=True)
product_line_count    

City      Product line          
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Name: proportion, dtype: float64